ui/apps/platform/src/Containers/NetworkGraph/types/topology.type.ts

```diff
@@ -81,11 +84,52 @@ export type NamespaceData = {
 
 export type NetworkPolicyState = 'none' | 'both' | 'ingress' | 'egress';
 
+// prettier-ignore
+type NodeModelType<DataType extends NodeDataType> =
+    DataType extends 'DEPLOYMENT' ? DeploymentNodeModel :
+    DataType extends 'INTERNAL_GROUP' ? InternalGroupNodeModel :
+    DataType extends 'EXTERNAL_GROUP' ? ExternalGroupNodeModel :
+    DataType extends 'EXTERNAL_ENTITIES' ? ExternalEntitiesNodeModel :
+    DataType extends 'CIDR_BLOCK' ? CIDRBlockNodeModel :
+    DataType extends 'UKNOWN_INTERNAL_ENTITY' ? UnknownInternalEntityNodeModel :
+    DataType extends 'INTERNAL_ENTITIES' ? InternalEntitiesNodeModel :
+    DataType extends 'EXTRANEOUS' ? ExtraneousNodeModel :
+    never;
+
+/**
+ * Returns a type guard for checking if a node is of a certain type
+ *
+ * @template DataType the type of node to check, should not be specified explicitly
+ * @param type the expected string value of the node's `data.type` property
+ * @returns a type guard for checking if a node is of a certain type
+ */
+export function isOfType<DataType extends NodeDataType>(
+    type: DataType
+): (node: CustomNodeModel) => node is NodeModelType<DataType> {
+    return (node: CustomNodeModel): node is NodeModelType<DataType> => node.data.type === type;
+}
+
+/**
+ * Type guard for checking if a node is of a certain type
+ *
+ * @template DataType the type of node to check, should not be specified explicitly
+ * @param type the expected string value of the node's `data.type` property
+ * @param node the node to check
+ * @returns true if the node is of the specified type, false otherwise
+ */
+export function isNodeOfType<DataType extends NodeDataType>(
+    type: DataType,
+    node: CustomNodeModel
+): node is NodeModelType<DataType> {
+    return isOfType(type)(node);
+}
+
```

Comment:
These functions serve the same purpose as the `node.data.type === 'SOME_TYPE'` checks but with the following benefits:

- Using these functions causes the correct node type to be inferred inside the `if` block without the need for casting with `as`
- Keeping a mapping of strings to node types encapsulated here prevents possible errors where `node.data.type` checks and `as NodeType` assertions are used, potentially out of sync
- A 2-arity form of `isNodeOfType(type, node)` and curried `isOfType(type)` form allows us to use these functions easily for checking a single node, or passed to a `filter` for type-safe filtering.